### PR TITLE
add new module: gcal

### PR DESCRIPTION
**Pull request by lasers**
`gcal`: *Display upcoming calendar events from Google Calendar* 

* Requires `gcalcli`

### CODE REVIEW

File: py3status/modules/gcal.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# -*- coding: utf-8 -*-
+"""
+Display upcoming calendar events from Google Calendar.
+
+Configuration parameters:
+    agenda_days: show calendar events for number of days (default 1)
```
Comment (tobes) [inline py3status/modules/gcal.py]: something like `days_shown` feels more clear.  I'm sure a better name exists.

File: py3status/modules/gcal.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    button_refresh: mouse button to refresh this module (default 2)
+    cache_timeout: refresh interval for this module (default 21600)
+    format: display format for this module (default '{format_event}')
+    format_event: display format for events
+        (default '[\?if=is_date&color=degraded %a %b %d ]
+        [\?if=is_time&color=good %-I:%M %p ][\?color=bad {event}]')
```
Comment (tobes) [inline py3status/modules/gcal.py]: why is the date `degraded` it feels slightly wrong to me.  I realise you want colourful.

File: py3status/modules/gcal.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    format_separator: show separator only if more than one (default ', ')
+    remove: list of strings to be removed from the output (default [])
+
+Format placeholders:
+    {format_event} format for events
+    is_event: a boolean based on available events
```
Comment (tobes) [inline py3status/modules/gcal.py]: hmm, I don't really like these here but not sure how we can deal with them better.  This is something that would be good to resolve.

File: py3status/modules/gcal.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+format_event placeholders:
+    {event} scheduled event eg Pizza night
+    %?  strftime characters to be translated. See `man strftime'.
```
Comment (tobes) [inline py3status/modules/gcal.py]: the last quote should be a ` not a '

File: py3status/modules/gcal.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+"""
+
+from datetime import datetime as dt, timedelta
+import re
+
+STRING_NOT_INSTALLED = "isn't installed"
+FMT_DATE = '%a %b %d'
+FMT_GCAL = '%Y%m%d'
+FMT_SS = '%s %s'
+FMT_TIME = '%I:%M%p'
```
Comment (tobes) [inline py3status/modules/gcal.py]: `%R` feels better it is better understood internationally imho

also the names are not very easy to understand `FORMAT_TIME` is better

File: py3status/modules/gcal.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    agenda_days = 1
+    button_refresh = 2
+    cache_timeout = 21600
```
Comment (tobes) [inline py3status/modules/gcal.py]: why this magic number?  Also this feels wrong.  If I have a calendar then I'd expect it to update more cleverly, but that can be added later.

File: py3status/modules/gcal.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    cache_timeout: refresh interval for this module (default 21600)
+    format: display format for this module (default '{format_event}')
+    format_event: display format for events
+        (default '[\?if=is_date&color=degraded %a %b %d ]
+        [\?if=is_time&color=good %-I:%M %p ][\?color=bad {event}]')
+    format_separator: show separator only if more than one (default ', ')
+    remove: list of strings to be removed from the output (default [])
+
+Format placeholders:
+    {format_event} format for events
```
Comment (tobes) [inline py3status/modules/gcal.py]: `{format_events}`?

File: py3status/modules/gcal.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            if out.rstrip():
+                temporary['date'] = out[:10].strip()
+                temporary['time'] = out[10:20].strip()
+                temporary['event'] = out[20:].strip()
+                temporary = {k: v for k, v in temporary.items() if v}
+                new_data.append(temporary)
+
+        return is_event, new_data
+
+    def _manipulate_data(self, data):
```
Comment (tobes) [inline py3status/modules/gcal.py]: this is a good example of where comments are useful.  What are we doing and why.  example data would also help understand what is going on.

File: py3status/modules/gcal.py (hunk 6 above)
Comment (lasers) [inline py3status/modules/gcal.py]: Six hours. I can't decide this part... What's the approximate `cache_timeout` for everybody? If we set dates long ago and/or use recurring events, then it's probably not an issue. Otherwise, we may connect with Google Calendar (via `gcalcli`) every `xx` seconds for little or no gain.

I think if I set it 6 hours, then the users could put in new number for `cache_timeout` if they need it to be updated more frequently. This module is pretty new so I didn't play with it much yet...What happen if I set `cache_timeout = 60`... and hadn't left home for an appointment... Would that `event` be on my status or just disappear right away if it's past `60` seconds? People might like the event to linger a bit longer on their status.. Will experiment this.

People might like having events there from few hours ago. `Bar with friends` on yours? 🍺 🍻

File: py3status/modules/gcal.py (hunk 8 above)
Comment (lasers) [inline py3status/modules/gcal.py]: gcal

We parse this (w/o colors). No `%R` for time... We turn them into `datetime objects` then we make them available for `strftime`. Every line is an event so we have `is_time` and `is_date` boolean checking in `format_event` because we can have... 
* `date`, `time`, `event`
* `date`, `____`, `event`
* `____`, `time`, `event`
* `____`, `____`, `event`

I hope that clears things up. Please ask away more questions. Thx. Cheers. 🍻

File: py3status/modules/gcal.py (hunk 1 above)
Comment (tobes) [inline py3status/modules/gcal.py]: > We're using command gcalcli agenda [start] [end] so I went with agenda_days

understandable but I'm not sure the user will pick up on it.  So I still think a better name is helpful

File: py3status/modules/gcal.py (hunk 2 above)
Comment (tobes) [inline py3status/modules/gcal.py]: > color=green

I have a branch that does this but haven't made a PR

Anyhow this does not address the issue that I raised.  Personally I think simple defaults is more important than colorful. at least if you used `color_date` it would make more sense to the user

File: py3status/modules/gcal.py (hunk 5 above)
Comment (tobes) [inline py3status/modules/gcal.py]: ok so `GCALCLI_TIME_FORMAT` might make sense then.  

> I'm just explaining things.

That's what comments and docstrings is for :D

File: py3status/modules/gcal.py (hunk 6 above)
Comment (tobes) [inline py3status/modules/gcal.py]: > People might like having events there from few hours ago. 

there are better ways to achieve this if it is wanted.  With as is now the time depends on when the refresh happens which is super inconsistent.

File: py3status/modules/gcal.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    {'color': '#00FF00', 'full_text': u'6:00 AM '},
+    {'color': '#FF0000', 'full_text': u'Garbage Pickup'},
+    {'full_text': u', '}
+    {'color': '#FFFF00', 'full_text': u'Sat Jul 21 '},
+    {'color': '#00FF00', 'full_text': u'6:05 PM '},
+    {'color': '#FF0000', 'full_text': u'Flight to Poland'}
+]
+"""
+
+from datetime import datetime as dt, timedelta
```
Comment (tobes) [inline py3status/modules/gcal.py]: `from datetime import datetime, timedelta` feels cleaner `dt` confuses me.

**Comment by lasers** [comment]
* I still like `agenda_days` over `days_shown`. Same as `forecast_days` in weather module(s).
* Removed `as dt. We use `datetime` for readability.
* Removed `is_event`. We can use `{format_event}|No Events` instead.
* We go with `1` hour update instead of `6` hours update.
* `FMT` to `FORMAT` for readability.
* It's possible to get `gcalcli` to print miltary time then parse them. Would have to add additional argument to `command` and change few things. I prefer to parse gcalcli's default output.
* There should be a bug at end of December because there are no years to parse... so I'm assuming today's year for both `[start]` and `[end]`. I opened an issue about that too.
* Instead of using `color_*`, the users can change the colors (as well as datetime styling) in `format`.

All done. Find anything else? Lemme know. Thx.

File: py3status/modules/gcal.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    button_refresh: mouse button to refresh this module (default 2)
+    cache_timeout: refresh interval for this module (default 3600)
+    command: customize gcalcli to dictate what to do with the events
+        (default 'gcalcli agenda --nocolor')
+    format: display format for this module (default '{format_event}')
+    format_event: display format for events
+        (default '[\?if=is_date&color=#ff0 %a %b %d ]
+        [\?if=is_time&color=#0f0 %-I:%M %p ][\?color=#f00 {event}]')
+    format_separator: show separator only if more than one (default ', ')
+    remove: list of strings to remove from the output (default [])
```
Comment (lasers) [inline py3status/modules/gcal.py]: @tobes. We could remove this option in favor of global sed-like option.

I added this so the users could `remove = [':00']` turning `7:00 AM` to `7 AM` since it is not possible to do `[\?if=is_time&color=#0f0 %-I[\?has_zero :%M] %p]` due to lack of placeholders. We still keep `%M` for `7:15 AM`, `7:30 AM`, `7:45 AM`, etc.

Comment (tobes) [inline py3status/modules/gcal.py]: I still very strongly think that these formats are too complex/crazy for defaults, and are actively anti user.

`color=#0f0` this type of thing is just evil.  We should not be hard coding a colour in a format as it makes it a real pain to remove.  If people want colourful output they are welcome to it but forcing it on people feels wrong.

on the 7 AM issue to me this is crazy if the time shows as `7:15 AM` then I should also see `7:00 AM` this is about consistency which is important.

I still think 24hr time should be used in the default output as that is more technical and wider used/understood.

File: py3status/modules/gcal.py (hunk 1 above)
Comment (tobes) [inline py3status/modules/gcal.py]: this name is still terrible imho

File: py3status/modules/gcal.py (hunk 10 above)
Comment (tobes) [inline py3status/modules/gcal.py]: can we just kill this?

File: py3status/modules/gcal.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+]
+"""
+
+from datetime import datetime, timedelta
+import re
+
+STRING_NOT_INSTALLED = "isn't installed"
+FORMAT_GCAL = '%Y%m%d'
+FORMAT_DATE = '%a %b %d'
+FORMAT_TIME = '%I:%M%p'
```
Comment (tobes) [inline py3status/modules/gcal.py]: These are still confusingly named as this is the format for calling gcalcli rather than outputting from the module.

Clarity is good for all the family.

File: py3status/modules/gcal.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    agenda_days = 1
+    button_refresh = 2
+    cache_timeout = 3600
+    command = 'gcalcli agenda --nocolor'
```
Comment (tobes) [inline py3status/modules/gcal.py]: why is this here?  what is it adding to the module?

File: py3status/modules/gcal.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    {'color': '#FFFF00', 'full_text': u'Sat Jul 22 '},
+    {'color': '#00FF00', 'full_text': u'6:05 PM '},
+    {'color': '#FF0000', 'full_text': u'Flight to Poland'}
+]
+"""
+
+from datetime import datetime, timedelta
+import re
+
+STRING_NOT_INSTALLED = "isn't installed"
```
Comment (tobes) [inline py3status/modules/gcal.py]: This doesn't feel to be very informative to a user.

File: py3status/modules/gcal.py (hunk 12 above)
Comment (lasers) [inline py3status/modules/gcal.py]: Some options that could be tweaked. http://sprunge.us/hiKZ 
Display shared events? Hide allday events? Import config from file? Etc.

File: py3status/modules/gcal.py (hunk 13 above)
Comment (lasers) [inline py3status/modules/gcal.py]: I forget `gcal: isn't installed`. It is now `gcal: gcalcli isn't installed`.

File: py3status/modules/gcal.py (hunk 10 above)
Comment (lasers) [inline py3status/modules/gcal.py]: >If people want colourful output they are welcome to it but forcing it on people feels
>wrong.

Many modules comes forced with `GOOD/BAD/DEGRADED` colors.

I'm not trying to force colors on people. This is to set up something sensible for users to see in default module. I would set up `darkyellow, darkcyan` just like default `gcalcli` output, but I feel that'd go against py3status to set random colors.

Color problem: There are more to things than just `GOOD/BAD/DEGRADED`.

Also, users can see where they can change colors in the format_event. Similar to
sysdata`s `format = [\?color=mem Mem...]`. Now I have to add examples. :-(

>on the 7 AM issue to me this is crazy if the time shows as 7:15 AM then I should also see 7:00 AM this is about consistency which is important.

This is consistency if people quickly add events: "Dinner at 6pm today", "Dentist at 8am today",
"Lunch with Joey at 11am today", "Oil change appointment at 3pm tomorrow", "Basketball
game at 8pm next Thursday"... It all would translates to `X:00 AM/PM` (depending on datetime
styling). Often, we don't need that `:00`. 

>can we just kill this?

I use `remove = [':00']` myself, but wanted an global option so this simple option could be killed and could be useful in other modules.

**Comment by lasers** [comment]
Example output for @tobes to work with:
```

Tue Aug 01           Volleyball

Thu Aug 03   6:00am  Garbage Pickup

Fri Aug 04   9:00am  Pickup

Sat Aug 05           Volleyball
                     Dropoff
             3:00pm  Fishing

```

**Comment by ultrabug** [comment]
As a G Suite user at work, I'm very interested in this module.

I tried it and gcalcli is a mess and looks abandoned. It even has encoding issues on the latest (alpha) release so tbh I'm not sure where to stand on this.

Maybe using the https://github.com/google/gdata-python-client/blob/master/samples/calendar/calendarExample.py example as a base to read only operations would be better ?

**Comment by lasers** [comment]
```
2d [lasers:~/Downloads] 2 % python2 ./calendarExample.py --user _____ --pw _____
Traceback (most recent call last):
  File "./calendarExample.py", line 617, in <module>
    main()
  File "./calendarExample.py", line 613, in main
    sample = CalendarExample(user, pw)
  File "./calendarExample.py", line 51, in __init__
    self.cal_client.ClientLogin(email, password, self.cal_client.source);
  File "/usr/lib/python2.7/site-packages/gdata/client.py", line 441, in client_login
    captcha_token=captcha_token, captcha_response=captcha_response)
  File "/usr/lib/python2.7/site-packages/gdata/client.py", line 373, in request_client_login_token
    response, ClientLoginFailed, response_body)
gdata.client.ClientLoginFailed: Server responded to ClientLogin request: 404, https://developers.google.com/accounts/docs/AuthForInstalledApps
```
https://developers.google.com/accounts/docs/AuthForInstalledApps

**Comment by ultrabug** [comment]
sux... just try with special chars like 'é' in your agenda, everything breaks like this for me:

```
➜  py3status python gcal.py
Traceback (most recent call last):
  File "gcal.py", line 165, in <module>
    module_test(Py3status, config=config)
  File "/home/alexys/github/py3status/py3status/module_test.py", line 52, in module_test
    print(getattr(module, method)(*method_args))
  File "gcal.py", line 138, in gcal
    data = self._get_gcal_data()
  File "gcal.py", line 90, in _get_gcal_data
    return self.py3.command_output(self.command % (start, end), shell=False)
  File "/home/alexys/github/py3status/py3status/py3.py", line 896, in command_output
    msg = msg.format(cmd=command[0], error=retcode, output=output)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xfb' in position 8: ordinal not in range(128)
```

**Comment by lasers** [comment]
Fixed a `list-was-in-wrong-place` bug. That special `e` works okay here. Let me know if it's still causing you grief then I'll keep trying things. Thx. :-) 
2017-08-01-020815_381x54_scrot

**Comment by ultrabug** [comment]
Same problem still, but I'm running gcalcli 4.0.0a3 because old versions are bugged due to the usage of old oauth2 python lib...

**Comment by lasers** [comment]
`gcalcli v3.4.0 (Eric Davis, Brian Hartvigsen)` here.

**Comment by tobes** [comment]
yeah maybe using the google official python client might be more sensible

**Comment by ultrabug** [comment]
Aye, I have doubts that we'll make this working with gcalcli clearly :(

**Comment by lasers** [comment]
I take it you have no problem running `gcalcli agenda`? 
>Aye, I have doubts that we'll make this working with gcalcli clearly :(

It works here. 🤣 Is this more of a `self.py3.command_output` issue? The error snippet above on that line `_get_gcal_data` is the line where we are only trying to get the data.
```
  File "gcal.py", line 90, in _get_gcal_data
    return self.py3.command_output(self.command % (start, end), shell=False)
  File "/home/alexys/github/py3status/py3status/py3.py", line 896, in command_output
```

**Comment by ultrabug** [comment]
I can run it alright from the command line yes. No clue why it fails that horribly afterwards (the reporting of the command_output is not very helpful either)

**Comment by ultrabug** [comment]
```
error Traceback (most recent call last):
  File "/usr/lib/python-exec/python2.7/gcalcli", line 2500, in <module>
    BowChickaWowWow()
  File "/usr/lib/python-exec/python2.7/gcalcli", line 2398, in BowChickaWowWow
    gcal.AgendaQuery(startText=FLAGS.start, endText=FLAGS.end)
  File "/usr/lib/python-exec/python2.7/gcalcli", line 1696, in AgendaQuery
    self._DisplayQueriedEvents(start, end)
  File "/usr/lib/python-exec/python2.7/gcalcli", line 1661, in _DisplayQueriedEvents
    self._IterateEvents(start, eventList, yearDate=yearDate)
  File "/usr/lib/python-exec/python2.7/gcalcli", line 1527, in _IterateEvents
    self._PrintEvent(event, prefix)
  File "/usr/lib/python-exec/python2.7/gcalcli", line 1213, in _PrintEvent
    PrintMsg(eventColor, fmt % ('', _u(self._ValidTitle(event).strip())))
  File "/usr/lib/python-exec/python2.7/gcalcli", line 410, in PrintMsg
    sys.stdout.write(msg)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 21: ordinal not in range(128)
```

**Comment by lasers** [comment]
Let me get you clearly... It's an issue with `gcalcli` + unicode and...
that was the output for `gcalcli agenda 2017080120170802`? 

Any error with `gcalcli agenda --tsv 2017080120170802`?

**Comment by lasers** [comment]
Changes being made...
* Use `Tab Separated Values` instead of parsing `agenda` output... An option I overlooked.
    This fixes the yearly bug since it printed everything I need.

* Add `is_multiple` indicator... eg `Vacation*`
    An asterisk next to an event implies that it's scheduled for more than a day.

* No issues here right now in this commit. This is ready for review and/or merge.

* Things I'd like to work on... eventually
    An `reminder` option (default 5min) + `urgent`.
    Cache offset and/or loop issues... mainly for reminding and urgent.... all without touching external `get_gcal_data` calls... If a first event is found to be less than `cache_timeout`, then we should use that one (minus reminder secs). Otherwise, back to `cache_timeout`. This sounds a bit tricky atm.

* Knowing that this might not work with @ultrabug's `gcalcli` right now. It shouldn't be a deal breaker since it's working okay for me here. Others can try this at their own perils. Feedback welcome. Ty.

**Comment by lasers** [comment]
As you may notice, there is another Google Calendar module now... I'm working new changes and/or features into this module to make things more useful / exciting for the users.

It's all in good competition... and all in good fun too. 🥇 🥈 🥉 

Changes made recently...

* Add buttons to focus next/previous event, open event links, reset (eg clears out the scrolling), and refresh. We don't call Google API during any of those... except for refresh. Very nice.

* Add arguments to dictate the events... No more exposing the command.
    1) Don't show events that have started
    2) Don't show events that have been declined
    `arguments = ['nostarted', 'nodeclined']`

* By using `--details=all` too, we have more new placeholders.
* I `**e` everything in the dict too. Slightly more (useless) placeholders. Less lines too.

* As mentioned in earlier comment, we are already using `Tab Separated Values`.

* This module was initially simple as it does not expose end dates/times at all.

   1) Showing start date/time is more than enough for the users.
    2) I now added them. They need to be in their own formats for strftime to work ... The alternative is to use separate fancy separators to parse between them in single `format_event`.
   3) Users can show `1:00 AM - 3:00 AM`. They can modify things too... Showing `1 - 3` (via `remove = [':00 AM', ':00 PM']` option). Google have this option to trim an hour event.. so you can have `1 - 1:50 PM` on the bar still... Nice.

* When you scroll, you can use `is_focused` to expose additional information such as `{description}` (useful), `{location}` (possibly useful in some cases), creator `{email}` (meh), and `{calendar}` owner (meh). 

* No issues here right now in this commit. This is ready for review and/or merge.

Things I'd like to work on... eventually
* An `reminder` option (default 15min) + `urgent`.
* Cache offset and/or loop issues... mainly for reminding and urgent.... all without touching external `get_gcal_data` calls... If a first event is found to be less than `cache_timeout`, then we should use that one (minus reminder secs). Otherwise, back to `cache_timeout`.

Feedback welcome. Ty.

**Comment by igorgrebenkov** [comment]
I just gave this a try. Works very well for me.

Grouping events by day is a good idea. I like the focus feature, too. Lots and lots of config parameters! There's lot of customization potential here. Good job. 

Is there a way to not have it show the weather for each day?

**Comment by lasers** [comment]
>Works very well for me.

Good. That translates to no bugs! :-)

>Grouping events by day is a good idea

It's a nice default. If you want to, you can modify `format_event_start` to show everything. Guessing without checking... (config below). It could be possible to add `is_toggled` inside `format_event` too then it could behave somewhat similar to your module where the events can be exposed and be left alone when scrolling through... Idk.
```
    format_event_start = '%a %b %-d %-I:%M %p'
```
>Is there a way to not have it show the weather for each day?

I exposed all placeholders here from command `gcalcli`... and command `calcli` does not give me weather data. What about yours? I'm sure you obtain more data than I do here. Weather too?

Also, we have `weather_yahoo` and very recently merged `weather_owm`.

**Comment by igorgrebenkov** [comment]
Yes, I use weather_yahoo currently, which is why it's superfluous for me in gcal. I figured out the weather thing. There's a setting in google calendar to display weather forecasts for each day. I had it off, but it was still fetching weather. I removed my location from the settings and it stopped doing that.

**Comment by lasers** [comment]
Interesting! I misread you first time as I never ran into this issue... and I recently tried this... I'm not able to reproduce your issue. Can you paste the output for `gcalcli agenda --tsv --details=all 2017-10-22 2017-10-23`? I'm thinking something might includes it as a part of event `{description}`.

**Comment by igorgrebenkov** [comment]
With a location set in Google Calendar settings:

`2017-10-22	00:00	2017-10-23	00:00	https://www.google.com/calendar/event?eid=MjAxNzEwMjJfd2VhdGhlclAgcCN3ZWF0aGVyQHY		Forecast for Ottawa, Canada (22° | 8°)			Weather: Milwaukee, WI	p#weather@group.v.calendar.google.com`

(There are no events on that day, but the forecast shows up as an event if I have a location set in Google Calendar settings)

**Comment by lasers** [comment]
@igorgrebenkov Can you try it now? I reject events with email containing `#weather`.

**Comment by igorgrebenkov** [comment]
That works, no more weather.

File: tests/test_module_doc.py [hunk 14]
```diff
@@ -35,6 +35,8 @@
 # Ignored items will not have their default values checked or be included for
 # alphabetical order purposes
 IGNORE_ITEM = [
+    ('gcal', 'format_event'),  # line too long for docstring parsing
+    ('gcal', 'format_event_start'),  # line too long for docstring parsing
```
Comment (lasers) [inline tests/test_module_doc.py]: @igorgrebenkov

Comment (igorgrebenkov) [reply tests/test_module_doc.py]: Great thanks!

**Comment by lasers** [comment]
@tobes I added `ignore` config to deal with recurring events (see `weather` comment above) that may looks pretty on the website, but not on the bar. Can you review this part? I don't feel I did this part nicely... (eg users could want more than 1 `event` items to ignore).

**Comment by tobes** [comment]
We now have #1068 and #961 that both add google calendar support I know @ultrabug is keen to add such a module I'm not sure which we will end up going with but I think it is best to get a basic one in and then improve the functionality.  I know that this is something I will be using for work so I have some specific features that I will want.

So I think this is a slowly slowly thing.  If you add something contentious then it end up blocking the whole PR.  (this can be a simple as a blank line in the wrong place as you know.

**Comment by lasers** [comment]
>I'm not sure which we will end up going with.

Easy! This one.... Just kidding... Both? Like weather, bitcoin, etc...

I assume you didn't play with either module. I'll make the pro and con list. @igorgrebenkov: You're welcome to help out with this list as you're the other author.

Notifications - `google_calendar` 🍏  :: `gcal` 🍎 
Open URLs - `google_calendar` 🍏  :: `gcal` 🍏 
Organized - `google_calendar` 🍎 :: `gcal` 🍏 
Polished - `google_calendar` 🍎 :: `gcal` 🍏 
UI Style - `google_calendar` clicking :: `gcal` scrolling then clicking
View Details  - `google_calendar` Expand all events :: `gcal` Expand one
Both requies hard dependencies. 🍎 :: 🍎 
Timers - `google_calendar` 🍏 :: `gcal`  🍎
Customzability - `google_calendar` ? (learning) :: `gcal` 🍏 (grouping events by day, multiple day, etc)
Display events based on number of... `google_calendar` events :: `gcal` days
Symbols  - `google_calendar` 🍎 :: `gcal`  🍏
Thresholds - `google_calendar` 🍏 :: `gcal` 🍎
Documentation - `google_calendar` 🍏 :: `gcal` 🍎

**Comment by tobes** [comment]
> I assume you didn't play with either module.

Yes this is correct

> I'll make the pro and con list. 

That's good but the layout is a little had to read - as a table is easier

Feature |google_calendar | Gcal
------------ | ------------- | -------------
Notifications | :green_apple:| :apple:
Open URLs| :green_apple:| :green_apple:

**Comment by igorgrebenkov** [comment]
I figured there might be room for both since they use different approaches...gcalcli vs an api key? I can imagine some people will prefer one approach or the other. That's up to you guys, though. That table sums it up pretty well so far, @lasers. After the color features I added in mine last night, I'm done with new features and will focus on fixing outstanding issues.

**Comment by ultrabug** [comment]
I think we should only have one module doing that and I think that having it pure python just like @igorgrebenkov google_calendar is the way to go.

This means that I would be delighted to see the cool features of @lasers gcal module ported to yours @igorgrebenkov !

This module has a huge potential and I've been dreaming about it for so long that I think it's worth working hard on it!

**Comment by lasers** [comment]
>I think we should only have one module...

😢 😭 

`backlight` does not work on some laptops... `light` to the rescue.
`mpris` is not perfect and/or is limited sometimes.... Dedicated `moc` + `cmus` to the rescue... both supporting 20~ placeholders easily. I could work on updating the others too.

`gpmdp`, `clementine`, `mpd`, etc could be rewritten from scratch. We cannot enrich the modules without starting the `mergeformat` process... and if applicable, `thresholds` too.

Praise the Lord for `scratchpad_counter` and `window_title`... 👼... because..... both `scratchpad_async` and `window_title_async` still causes issues here _(as well as `window_manipulation` 💨)._ I kept seeing `py3status` pleading `100%` temporary insanity 💩 .... on `htop`.... repeatedly. I have to kill `py3status` to regain normalacy.

We brought `deadbeef` back from dead #844 along with the cool extras! #876.

Someday, we can make `opensuse_updates` and maybe update `fedora_updates` too... and see if we're missing some `*_updates` for popular distros. Centos? Gentoo? [List_of_Linux_distributions](https://en.wikipedia.org/wiki/List_of_Linux_distributions) 

After this, see if `py3status` can go into linux distros repositories.

Also, I have several `newmodule-*` branches to work on... writing... making... fixing... testing... polishing... optimizing... et cetera.... all for `py3status`. :-)

I'd prefer to keep `gcal` as I currently use it... and with `gcalcli`,  we can edit events using the terminal (cli) rather than the website... So that's one 🍏 for `gcal`. _(Perfect for quick add <event>)._ I'll leave it up to @igorgrebenkov (and others) to port delicious features over to `google_calendar`. 👅 

There are other sitting PRs too. What to do with them? 🎃

**Comment by igorgrebenkov** [comment]
Sorry, @lasers didn't mean to steal your thunder here. I'll be happy to help port over cool features from this module!

**Comment by lasers** [comment]
⚡️ @igorgrebenkov ⚡️.... `gcalcli` hadn't been updated for >1 year. You're good. 👍

**Comment by tobes** [comment]
I think we now have the `google_calendar` module so can this now be closed?

**Comment by lasers** [comment]
One day, I want to replace this with [khal](https://github.com/pimutils/khal) one day.

**Comment by tobes** [comment]
I'd prefer a more generic calendar module than just a google specific one at some point but maybe it is good to get that one working nicely and then allowing other calendar sources

**Comment by lasers** [comment]
That's what `khal` should be... a generic calendar tool... I just need to understand it more. `khal` using `vdirsyncer` would mean working with any other sources too... so this would be something we want to implement.... and not worrying about anything else. 👍

**Comment by ultrabug** [comment]
Okay, wishes are good but I think we can close this PR now?

**Comment by lasers** [comment]
Sure? I mean, I can rename/make new commits here no problem. I wanted to get lot of  other PRs out of way first before I started trying on this... I should only have to replace `_organize` method, then update names... and well, just glue things together.

I want to know what to do with `thunderbird_mail` PR too.

**Comment by ultrabug** [comment]
If so @lasers then I think it's worth a new PR to start fresh later. Thanks again!
